File: jvm/selfie-lib/src/commonTest/kotlin/com/diffplug/selfie/guts/EscapeLeadingWhitespaceTest.kt
/*
 * Copyright (C) 2024 DiffPlug
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.diffplug.selfie.guts

import com.diffplug.selfie.guts.EscapeLeadingWhitespace.*
import com.diffplug.selfie.guts.EscapeLeadingWhitespace.Companion.appropriateFor
import io.kotest.matchers.shouldBe
import kotlin.test.Test

class EscapeLeadingWhitespaceTest {
  @Test
  fun detection() {
    // not enough to detect
    appropriateFor("") shouldBe ALWAYS
    appropriateFor("abc") shouldBe ALWAYS
    appropriateFor("abc\nabc") shouldBe ALWAYS

    // all spaces -> only tabs need escape
    appropriateFor(" ") shouldBe ALWAYS
    appropriateFor("  ") shouldBe ONLY_ON_TAB
    appropriateFor("  \n  ") shouldBe ONLY_ON_TAB

    // all tabs -> only space needs escape
    appropriateFor("\t") shouldBe ONLY_ON_SPACE
    appropriateFor("\t\t") shouldBe ONLY_ON_SPACE
    appropriateFor("\t\n\t") shouldBe ONLY_ON_SPACE

    // it's a mess -> everything needs escape
    appropriateFor("\t\n  ") shouldBe ALWAYS

    // single spaces and tabs -> only tabs need escape
    appropriateFor(
        """
/*
${' '}* Copyright
${' '}*/
interface Foo {
${'\t'}fun bar()
}
""") shouldBe
        ONLY_ON_SPACE
  }
}

File: python/selfie-lib/tests/EscapeLeadingWhitespace_test.py
from selfie_lib import EscapeLeadingWhitespace


def test_detection():
    # not enough to detect
    assert EscapeLeadingWhitespace.appropriate_for("") == EscapeLeadingWhitespace.ALWAYS
    assert (
        EscapeLeadingWhitespace.appropriate_for("abc") == EscapeLeadingWhitespace.ALWAYS
    )
    assert (
        EscapeLeadingWhitespace.appropriate_for("abc\nabc")
        == EscapeLeadingWhitespace.ALWAYS
    )

    # all spaces -> only tabs need escape
    assert (
        EscapeLeadingWhitespace.appropriate_for(" ") == EscapeLeadingWhitespace.ALWAYS
    )
    assert (
        EscapeLeadingWhitespace.appropriate_for("  ")
        == EscapeLeadingWhitespace.ONLY_ON_TAB
    )
    assert (
        EscapeLeadingWhitespace.appropriate_for("  \n  ")
        == EscapeLeadingWhitespace.ONLY_ON_TAB
    )

    # all tabs -> only space needs escape
    assert (
        EscapeLeadingWhitespace.appropriate_for("\t")
        == EscapeLeadingWhitespace.ONLY_ON_SPACE
    )
    assert (
        EscapeLeadingWhitespace.appropriate_for("\t\t")
        == EscapeLeadingWhitespace.ONLY_ON_SPACE
    )
    assert (
        EscapeLeadingWhitespace.appropriate_for("\t\n\t")
        == EscapeLeadingWhitespace.ONLY_ON_SPACE
    )

    # it's a mess -> everything needs escape
    assert (
        EscapeLeadingWhitespace.appropriate_for("\t\n  ")
        == EscapeLeadingWhitespace.ALWAYS
    )

    # single spaces and tabs -> only tabs need escape
    test_string = f"""/*
 * Copyright
 */
interface Foo [
{'\t'} bar()
]"""
    assert (
        EscapeLeadingWhitespace.appropriate_for(test_string)
        == EscapeLeadingWhitespace.ALWAYS
    )
